File: examples/php/complex/compiler_tests.php
<?php
/**
 * Compiler architecture test cases for the tree-sitter-sast flow engine.
 *
 * Each section tests a specific capability with both positive (flow expected)
 * and negative (no flow expected) scenarios.
 */

// ---------------------------------------------------------------------------
// 1. BRANCH-AWARE KILL ANALYSIS
// ---------------------------------------------------------------------------

// [NO FLOW EXPECTED] Both branches sanitize
function branch_kill_both_paths(string $user_input): void {
    if (strlen($user_input) > 100) {
        $cmd = "default_command";
    } else {
        $cmd = "safe_fallback";
    }
    system($cmd); // safe: both paths overwrite
}

// [FLOW EXPECTED] Only one branch kills taint
function branch_kill_one_path(string $user_input): void {
    $cmd = $user_input;
    if (strlen($cmd) > 100) {
        $cmd = "truncated";
    }
    // else: $cmd still holds $user_input
    system($cmd); // tainted
}


// ---------------------------------------------------------------------------
// 2. INTERPROCEDURAL KILL (2+ levels deep)
// ---------------------------------------------------------------------------

function sanitize_input(string $value): string {
    return str_replace(["'", ";", "--"], "", $value);
}

function wrap_sanitize(string $value): string {
    return sanitize_input($value);
}

// [NO FLOW EXPECTED] Sanitizer applied
function interprocedural_kill_direct(string $user_input): void {
    $clean = sanitize_input($user_input);
    $conn = new mysqli("localhost", "root", "", "test");
    $conn->query("SELECT * FROM t WHERE x = '$clean'");
}

// [NO FLOW EXPECTED] Sanitizer 2 levels deep
function interprocedural_kill_two_levels(string $user_input): void {
    $clean = wrap_sanitize($user_input);
    $conn = new mysqli("localhost", "root", "", "test");
    $conn->query("SELECT * FROM t WHERE x = '$clean'");
}

// [FLOW EXPECTED] No sanitizer
function interprocedural_no_kill(string $user_input): void {
    $conn = new mysqli("localhost", "root", "", "test");
    $conn->query("SELECT * FROM t WHERE x = '$user_input'");
}


// ---------------------------------------------------------------------------
// 3. FIELD-SENSITIVE TRACKING
// ---------------------------------------------------------------------------

class Config {
    public string $command = "";
    public string $label = "";
}

// [FLOW EXPECTED] Tainted field flows to sink
function field_sensitive_tainted(string $user_input): void {
    $cfg = new Config();
    $cfg->command = $user_input;
    $cfg->label = "safe_label";
    system($cfg->command); // tainted
}

// [NO FLOW EXPECTED] Only safe field used
function field_sensitive_safe(string $user_input): void {
    $cfg = new Config();
    $cfg->command = $user_input;
    $cfg->label = "safe_label";
    system($cfg->label); // safe
}

// [NO FLOW EXPECTED] Tainted field overwritten
function field_sensitive_overwrite(string $user_input): void {
    $cfg = new Config();
    $cfg->command = $user_input;
    $cfg->command = "hardcoded_safe";
    system($cfg->command); // safe: overwritten
}


// ---------------------------------------------------------------------------
// 4. EXCEPTION CHAIN TAINT
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] Taint flows through exception message
function exception_taint(string $user_input): void {
    try {
        throw new RuntimeException("Invalid: " . $user_input);
    } catch (RuntimeException $e) {
        system($e->getMessage()); // tainted
    }
}


// ---------------------------------------------------------------------------
// 5. SUPERGLOBAL SOURCE (PHP-specific)
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] $_GET is a taint source
function superglobal_source(): void {
    $name = $_GET['name'];
    system("echo $name"); // tainted
}

// [FLOW EXPECTED] $_POST source
function post_source(): void {
    $cmd = $_POST['command'];
    system($cmd); // tainted
}


// ---------------------------------------------------------------------------
// 6. FOREACH REFERENCE WRITE-BACK (PHP-specific)
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] Reference write-back taints array
function foreach_ref_writeback(string $user_input): void {
    $items = ["a", "b", "c"];
    foreach ($items as &$val) {
        $val = $user_input;
    }
    system($items[0]); // tainted: write-back modified array
}


// ---------------------------------------------------------------------------
// 7. TYPE-TRACKED CALL RESOLUTION
// ---------------------------------------------------------------------------

interface Executor {
    public function execute(string $cmd): void;
}

class SafeExecutor implements Executor {
    public function execute(string $cmd): void {
        echo "Would execute: $cmd\n"; // safe
    }
}

class UnsafeExecutor implements Executor {
    public function execute(string $cmd): void {
        system($cmd); // dangerous
    }
}

// [FLOW EXPECTED]
function type_tracked_unsafe(string $user_input): void {
    $executor = new UnsafeExecutor();
    $executor->execute($user_input);
}

// [NO FLOW EXPECTED]
function type_tracked_safe(string $user_input): void {
    $executor = new SafeExecutor();
    $executor->execute($user_input);
}


// ---------------------------------------------------------------------------
// 8. PARAMETER SOURCE SYNTHESIS
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] External entry point
function handle_webhook(string $payload, string $signature): void {
    system("process $payload");
}


// ---------------------------------------------------------------------------
// 9. CROSS-METHOD CLASS FIELD TAINT
// ---------------------------------------------------------------------------

class Pipeline {
    private string $data = "";

    public function ingest(string $raw): void {
        $this->data = $raw;
    }

    public function process(): void {
        system($this->data); // tainted if ingest() called with tainted input
    }
}

// [FLOW EXPECTED]
function cross_method_field_taint(string $user_input): void {
    $p = new Pipeline();
    $p->ingest($user_input);
    $p->process();
}


// ---------------------------------------------------------------------------
// 10. MATCH EXPRESSION (PHP 8)
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] Taint flows through match expression
function match_expression_taint(string $user_input): void {
    $action = match($user_input[0]) {
        'r' => "read $user_input",
        'w' => "write $user_input",
        default => $user_input,
    };
    system($action); // tainted in all branches
}


// ---------------------------------------------------------------------------
// 11. STRING INTERPOLATION
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] Complex interpolation
function string_interpolation_taint(string $user_input): void {
    $query = "SELECT * FROM users WHERE name = '{$user_input}'";
    $conn = new mysqli("localhost", "root", "", "test");
    $conn->query($query); // tainted
}


// ---------------------------------------------------------------------------
// 12. CONTAINER APPEND
// ---------------------------------------------------------------------------

// [FLOW EXPECTED] Taint flows through array_push
function container_append_tainted(string $user_input): void {
    $parts = [];
    array_push($parts, $user_input);
    $query = implode(" ", $parts);
    $conn = new mysqli("localhost", "root", "", "test");
    $conn->query($query);
}

File: examples/java/complex/CompilerTests.java
/**
 * Compiler architecture test cases for the tree-sitter-sast flow engine.
 *
 * Each section tests a specific capability with both positive (flow expected)
 * and negative (no flow expected) scenarios.
 */
import java.sql.*;
import java.util.*;

public class CompilerTests {

    // -----------------------------------------------------------------------
    // 1. BRANCH-AWARE KILL ANALYSIS
    // -----------------------------------------------------------------------

    // [NO FLOW EXPECTED] Both branches sanitize
    public void branchKillBothPaths(String userInput) throws Exception {
        String cmd;
        if (userInput.length() > 100) {
            cmd = "default_command";
        } else {
            cmd = "safe_fallback";
        }
        Runtime.getRuntime().exec(cmd); // safe
    }

    // [FLOW EXPECTED] Only one branch kills taint
    public void branchKillOnePath(String userInput) throws Exception {
        String cmd = userInput;
        if (cmd.length() > 100) {
            cmd = "truncated";
        }
        Runtime.getRuntime().exec(cmd); // tainted: else path preserves
    }

    // -----------------------------------------------------------------------
    // 2. INTERPROCEDURAL KILL (2+ levels deep)
    // -----------------------------------------------------------------------

    private String sanitize(String value) {
        return value.replaceAll("['\";\\-\\-]", "");
    }

    private String wrapSanitize(String value) {
        return sanitize(value);
    }

    // [NO FLOW EXPECTED] Sanitizer applied
    public void interproceduralKillDirect(String userInput) throws Exception {
        String clean = sanitize(userInput);
        Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:");
        conn.createStatement().execute("SELECT * FROM t WHERE x = '" + clean + "'");
    }

    // [NO FLOW EXPECTED] Sanitizer 2 levels deep
    public void interproceduralKillTwoLevels(String userInput) throws Exception {
        String clean = wrapSanitize(userInput);
        Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:");
        conn.createStatement().execute("SELECT * FROM t WHERE x = '" + clean + "'");
    }

    // [FLOW EXPECTED] No sanitizer
    public void interproceduralNoKill(String userInput) throws Exception {
        Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:");
        conn.createStatement().execute("SELECT * FROM t WHERE x = '" + userInput + "'");
    }

    // -----------------------------------------------------------------------
    // 3. FIELD-SENSITIVE TRACKING
    // -----------------------------------------------------------------------

    static class Config {
        String command = "";
        String label = "";
    }

    // [FLOW EXPECTED] Tainted field flows to sink
    public void fieldSensitiveTainted(String userInput) throws Exception {
        Config cfg = new Config();
        cfg.command = userInput;
        cfg.label = "safe_label";
        Runtime.getRuntime().exec(cfg.command);
    }

    // [NO FLOW EXPECTED] Only safe field used
    public void fieldSensitiveSafe(String userInput) throws Exception {
        Config cfg = new Config();
        cfg.command = userInput;
        cfg.label = "safe_label";
        Runtime.getRuntime().exec(cfg.label);
    }

    // [NO FLOW EXPECTED] Tainted field overwritten
    public void fieldSensitiveOverwrite(String userInput) throws Exception {
        Config cfg = new Config();
        cfg.command = userInput;
        cfg.command = "hardcoded_safe";
        Runtime.getRuntime().exec(cfg.command);
    }

    // -----------------------------------------------------------------------
    // 4. EXCEPTION CHAIN TAINT
    // -----------------------------------------------------------------------

    // [FLOW EXPECTED] Taint flows through exception message
    public void exceptionTaint(String userInput) throws Exception {
        try {
            throw new RuntimeException("Invalid: " + userInput);
        } catch (RuntimeException e) {
            Runtime.getRuntime().exec(e.getMessage());
        }
    }

    // -----------------------------------------------------------------------
    // 5. CONTAINER ELEMENT PRECISION
    // -----------------------------------------------------------------------

    // [FLOW EXPECTED] Tainted element added to list
    public void containerAppendTainted(String userInput) throws Exception {
        List<String> parts = new ArrayList<>();
        parts.add(userInput);
        String query = String.join(" ", parts);
        Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:");
        conn.createStatement().execute(query);
    }

    // -----------------------------------------------------------------------
    // 6. TYPE-TRACKED CALL RESOLUTION
    // -----------------------------------------------------------------------

    interface Executor {
        void execute(String cmd) throws Exception;
    }

    static class SafeExecutor implements Executor {
        public void execute(String cmd) {
            System.out.println("Would execute: " + cmd);
        }
    }

    static class UnsafeExecutor implements Executor {
        public void execute(String cmd) throws Exception {
            Runtime.getRuntime().exec(cmd);
        }
    }

    // [FLOW EXPECTED]
    public void typeTrackedUnsafe(String userInput) throws Exception {
        Executor executor = new UnsafeExecutor();
        executor.execute(userInput);
    }

    // [NO FLOW EXPECTED]
    public void typeTrackedSafe(String userInput) throws Exception {
        Executor executor = new SafeExecutor();
        executor.execute(userInput);
    }

    // -----------------------------------------------------------------------
    // 7. PARAMETER SOURCE SYNTHESIS
    // -----------------------------------------------------------------------

    // [FLOW EXPECTED] External entry point
    public void handleWebhook(String payload, String signature) throws Exception {
        Runtime.getRuntime().exec("process " + payload);
    }

    // -----------------------------------------------------------------------
    // 8. CROSS-METHOD CLASS FIELD TAINT
    // -----------------------------------------------------------------------

    static class Pipeline {
        private String data = "";

        void ingest(String raw) {
            this.data = raw;
        }

        void process() throws Exception {
            Runtime.getRuntime().exec(this.data);
        }
    }

    // [FLOW EXPECTED]
    public void crossMethodFieldTaint(String userInput) throws Exception {
        Pipeline p = new Pipeline();
        p.ingest(userInput);
        p.process();
    }

    // -----------------------------------------------------------------------
    // 9. STREAM API TAINT
    // -----------------------------------------------------------------------

    // [FLOW EXPECTED] Taint flows through stream operations
    public void streamTaint(String userInput) throws Exception {
        List<String> inputs = Arrays.asList(userInput, "safe");
        String result = inputs.stream()
            .filter(s -> s.length() > 0)
            .reduce("", (a, b) -> a + " " + b);
        Runtime.getRuntime().exec(result);
    }
}
